ambari-server/src/main/java/org/apache/ambari/server/utils/SecretReference.java

```diff
@@ -172,4 +231,56 @@ public static void replacePasswordsWithReferencesForCustomProperties(Map<String,
       }
     }
   }
+
+  /**
+   * Replaces all password type properties in the given {@link Configuration} object. Creates a new Configuration
+   * object instead of mutating the input configuration.
+   * @param configuration the input configuration
+   * @param passwordProperties password type properties in a multimap.
+   *                           It has {@code config-type -> [password-prop-1, password-prop-2, ...]} structure.
+   * @return a new configuration with password properties replaced
+   */
+  public static Configuration replacePasswordsInConfigurations(Configuration configuration,
+                                                                     Multimap<String, String> passwordProperties) {
+    // replace passwords in config properties
+    Map<String, Map<String, String>> replacedProperties = configuration.getProperties().entrySet().stream().map(
+      entry -> {
+        String configType = entry.getKey();
+        Map<String, String> replacedConfigProps =
+          replacePasswordsInPropertyMap(entry.getValue(), configType, passwordProperties);
+        return new SimpleEntry<>(configType, replacedConfigProps);
+      }).collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    // replace passwords in config attributes
+    Map<String, Map<String, Map<String, String>>> replacedAttributes = configuration.getAttributes().entrySet().stream().map(
+      configTypeEntry -> {
+        String configType = configTypeEntry.getKey();
+        Map<String, Map<String, String>> replacedConfigProps = configTypeEntry.getValue().entrySet().stream().collect(
+          toMap(Map.Entry::getKey, e -> replacePasswordsInPropertyMap(e.getValue(), configType, passwordProperties)));
+        return new SimpleEntry<>(configType, replacedConfigProps);
+      }).collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
```

Comment:
Can you please extract these lambda expressions to a function that can be referenced and tested more easily?

Reply:
These are simple lambdas too. Changed the layaout a bit to make the more readable.

Reply:
I am thinking on a simplification.